content/automate/ManageIQ/Cloud/VM/Provisioning/StateMachines/Methods.class/__methods__/amazon_customizerequest.rb

```diff
@@ -2,99 +2,166 @@
-        # Search for customization templates enabled for Cloud-Init that match the template/image name
-        if customization_template.blank?
-          log(:info, "Searching for customization templates (Cloud-Init) enabled that are named: #{customization_template_search_by_template_name}")
-          customization_template = prov.eligible_customization_templates.detect { |ct| ct.name.casecmp(customization_template_search_by_template_name)==0 }
-        end
-      end
-      unless customization_template_search_by_ws_values.nil?
-        # Search for customization templates enabled for Cloud-Init that match ws_values[:customization_template]
-        if customization_template.blank?
-          log(:info, "Searching for customization templates (Cloud-Init) enabled that are named: #{customization_template_search_by_ws_values}")
-          customization_template = prov.eligible_customization_templates.detect { |ct| ct.name.casecmp(customization_template_search_by_ws_values)==0 }
-        end
-      end
-      if customization_template.blank?
-        log(:warn, "Failed to find matching Customization Template", true)
-      else
-        log(:info, "Found Customization Template ID: #{customization_template.id} Name: #{customization_template.name} Description: #{customization_template.description}")
-        prov.set_customization_template(customization_template) rescue nil
-        log(:info, "Provisioning object updated {:customization_template_id => #{prov.get_option(:customization_template_id).inspect}}")
-        log(:info, "Provisioning object updated {:customization_template_script => #{prov.get_option(:customization_template_script).inspect}}")
-      end
-    else
-      log(:info, "Customization Template selected from dialog ID: #{prov.get_option(:customization_template_id).inspect}} Script: #{prov.get_option(:customization_template_script).inspect}")
-    end
-  end # case mapping
-  log(:info, "Processing Amazon customizations...Complete", true)
-end
+                def prov
+                  @prov ||= @handle.root["miq_provision"].tap do |req|
+                    raise 'miq_provision not specified' if req.nil?
+                  end
+                end
 
-def log(level, msg, update_message=false)
-  $evm.log(level, "#{msg}")
-  $evm.root['miq_provision'].message = "#{msg}" if $evm.root['miq_provision'] && update_message
-end
+                def template
+                  @template ||= prov.vm_template.tap do |req|
+                    raise 'vm_template not specified' if req.nil?
+                  end
+                end
+
+                def provider
+                  @provider ||= template.ext_management_system.tap do |ems|
+                    raise 'ext_management_system not specified' if ems.nil?
+                  end
+                end
+
+                def product
+                  @product ||= template.operating_system['product_name'].downcase rescue nil
```

Comment:
@pkomanek This refactoring, while it did not introduce the `rescue` statements, should be able to easily remove them.

For most the initial object is guaranteed to be valid because the accessor method is raising an error.  Here the `template` method will always return a `template` object or raise an error.  So we just need to worry about `operating_system` and the remaining items being nil.

If we use the `.` notation instead of `['product_name']` to reference the property you can use safe navigation to achieve this:
```ruby
@product ||= template.operating_system&.product_name&.downcase
```